pkg/controller/volume/attachdetach/attach_detach_controller_test.go

```diff
@@ -545,3 +552,87 @@ func volumeAttachmentRecoveryTestCase(t *testing.T, tc vaTest) {
 	}
 
 }
+
+func Test_VARecovery_CSIMigration(t *testing.T) {
+	fakeKubeClient := controllervolumetesting.CreateTestClient()
+	informerFactory := informers.NewSharedInformerFactory(fakeKubeClient, time.Second*1)
+	plugins := gcepd.ProbeVolumePlugins()
+	plugins = append(plugins, csi.ProbeVolumePlugins()...)
+
+	// Create the controller
+	adcObj, err := NewAttachDetachController(
+		fakeKubeClient,
+		informerFactory.Core().V1().Pods(),
+		informerFactory.Core().V1().Nodes(),
+		informerFactory.Core().V1().PersistentVolumeClaims(),
+		informerFactory.Core().V1().PersistentVolumes(),
+		informerFactory.Storage().V1().CSINodes(),
+		informerFactory.Storage().V1().CSIDrivers(),
+		informerFactory.Storage().V1().VolumeAttachments(),
+		nil, /* cloud */
+		plugins,
+		nil, /* prober */
+		false,
+		1*time.Second,
+		DefaultTimerConfig,
+		nil, /* filteredDialOptions */
+	)
+	if err != nil {
+		t.Fatalf("NewAttachDetachController failed with error. Expected: <no error> Actual: <%v>", err)
+	}
+	adc := adcObj.(*attachDetachController)
+
+	type testcase struct {
+		testName         string
+		volumeHandle     string
+		csiMigration     bool
+		expectVolumeName string
+	}
+	testcases := []testcase{
+		{
+			testName:         "csi migration is not enabled, should get the original volumeName",
+			csiMigration:     false,
+			volumeHandle:     "vol1",
+			expectVolumeName: "kubernetes.io/gce-pd/vol1",
+		},
+		{
+			testName:         "csi migration is enabled, should get the CSI translated volumeName",
+			csiMigration:     true,
+			volumeHandle:     "vol2",
+			expectVolumeName: "kubernetes.io/csi/pd.csi.storage.gke.io^projects/UNSPECIFIED/zones/UNSPECIFIED/disks/vol2",
+		},
+	}
+
+	for _, tc := range testcases {
+
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIMigration, tc.csiMigration)()
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIMigrationGCE, tc.csiMigration)()
+
+		pv := controllervolumetesting.NewPV("migrated-pv", tc.volumeHandle)
+		volumeSpec := volume.NewSpecFromPersistentVolume(pv, false)
+		plugin, _ := adc.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec)
+		pluginName := plugin.GetPluginName()
+		if pluginName != "kubernetes.io/gce-pd" {
+			t.Fatalf("Expect to get plugin \"kubernetes.io/gce-pd\", but got: %s", pluginName)
+		}
+
+		// suppose CSI migration is on for test plugin
```

Comment:
Should we copy the entire implementation from ad-controller in test? Seems to defeat the purpose of test. How about something like - https://github.com/gnufied/kubernetes/commit/8be54b48fe4f65276990d93fad2f957dbd6d817f#diff-47fb572faf452565e30c9a91b7e72693d2c6e4bcaecd43aaa9ccb5277b036ef3R385 ?

In this way we just verify `attachedVolumes` inside ASOW. The test should check if volume spec belongs to CSI driver and now intree driver.

Reply:
Done. I verified that the call of processVolumeAttachment will not mark the volume as uncertain with this fix, which indicates the fix works.